analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs

```diff
@@ -24,48 +24,74 @@ namespace SonarAnalyzer.Rules.CSharp
                     }
                 }
-            },
-            SyntaxKind.ClassDeclaration,
-            SyntaxKindEx.RecordClassDeclaration);
+            });
 
-        private static bool HasVirtualMembers(TypeDeclarationSyntax typeDeclaration)
-        {
-            var classMembers = typeDeclaration.Members;
-            return classMembers.OfType<MethodDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<PropertyDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<IndexerDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword))
-                   || classMembers.OfType<EventDeclarationSyntax>().Any(member => member.Modifiers.Any(SyntaxKind.VirtualKeyword));
-        }
+        private static bool HasVirtualMembers(TypeDeclarationSyntax typeDeclaration) =>
+            typeDeclaration.Members
+                .Where(member => member.IsAnyKind(PossiblyVirtualKinds))
+                .Any(member => member.Modifiers().Any(SyntaxKind.VirtualKeyword));
 
-        private static bool IsPrivateButNotSealedType(TypeDeclarationSyntax typeDeclaration) =>
-            typeDeclaration.Modifiers.Any(SyntaxKind.PrivateKeyword)
-            && !typeDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword)
+        private static bool IsNotSealed(TypeDeclarationSyntax typeDeclaration) =>
+            !typeDeclaration.Modifiers.Any(SyntaxKind.StaticKeyword)
             && !typeDeclaration.Modifiers.Any(SyntaxKind.SealedKeyword)
             && !typeDeclaration.Modifiers.Any(SyntaxKind.AbstractKeyword);
 
-        private static bool IsPrivateTypeInherited(INamedTypeSymbol privateTypeInfo) =>
-            privateTypeInfo.ContainingType.GetAllNamedTypes()
-                                          .Any(symbol => !symbol.MetadataName.Equals(privateTypeInfo.MetadataName)
-                                                         && symbol.DerivesFrom(privateTypeInfo));
+        private static bool IsPrivateOrFileTypeAndInherited(TypeDeclarationSyntax declaration, Lazy<SemanticModel> model, Lazy<List<INamedTypeSymbol>> otherSymbols)
+        {
+            if (declaration.Modifiers.Any(SyntaxKind.PrivateKeyword))
+            {
+                var symbol = model.Value.GetDeclaredSymbol(declaration);
+                return symbol.ContainingType.GetAllNamedTypes()
+                    .Any(other => !other.MetadataName.Equals(symbol.MetadataName) && other.DerivesFrom(symbol));
+            }
+            if (declaration.Modifiers.Any(SyntaxKindEx.FileKeyword))
+            {
+                var symbol = model.Value.GetDeclaredSymbol(declaration);
+                return otherSymbols.Value
+                    .Exists(other => !other.MetadataName.Equals(symbol.MetadataName) && other.DerivesFrom(symbol));
+            }
+
+            return true;
```

Comment:
Based on the method name, I was confused why it returns `true` here. I needed to reread the implementation and usage several times to understand and I am still a bit confused.
If possible, rename it to make it easier to understand.

Reply:
You are absolutely right, it's a terrible name.

I got a bit tunnel-visioned because I had the context, i renamed it to `IsPossiblyDerived`, meaning it's either private+derived, file-scoped+derived, or _possibly_ derived if it's not private/file-scoped.